network/src/lib.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Cumulus-specific network implementation.
+//!
+//! Contains message send between collators and logic to process them.
+
+use substrate_client::error::{Error as ClientError};
+use sr_primitives::traits::{Block as BlockT};
+use substrate_consensus_common::block_validation::{Validation, BlockAnnounceValidator};
+
+use polkadot_primitives::{
+	Hash as PHash, parachain::{CandidateReceipt, ValidatorIndex, ValidatorSignature, ValidatorId},
+};
+use polkadot_statement_table::Statement;
+use polkadot_validation::check_statement;
+
+use codec::{Decode, Encode};
+
+use std::marker::PhantomData;
+
+/// Justification that a parachain block is the parachain block candidate of one of the relay chain
+/// validator.
+#[derive(Encode, Decode)]
+pub struct BlockCandidateJustification {
+	/// Receipt of the parachain block candidate of the signer.
+	candidate_receipt: CandidateReceipt,
+	/// Signer of `signature`.
+	signer: ValidatorIndex,
+	/// Signature of the Candidate statement with `candidate_receipt`.
+	signature: ValidatorSignature,
+	/// The parent block of which the candidate must be include.
+	relay_chain_parent_hash: PHash,
+}
+
+/// Validate that data is a valid justification form a relay-chain validator that the block is a
+/// valid parachain-block candidate.
+pub struct JustifiedBlockAnnounceValidator<B> {
+	authorities: Vec<ValidatorId>,
+	phantom: PhantomData<B>,
+}
+
+impl<B: BlockT> JustifiedBlockAnnounceValidator<B> {
+	pub fn new(authorities: Vec<ValidatorId>) -> Self {
+		Self {
+			authorities,
+			phantom: Default::default(),
+		}
+	}
+}
+
+impl<B: BlockT> BlockAnnounceValidator<B> for JustifiedBlockAnnounceValidator<B> {
+	fn validate(&mut self, header: &B::Header, mut data: &[u8])
+		-> Result<Validation, Box<dyn std::error::Error + Send>>
+	{
+		let justification = BlockCandidateJustification::decode(&mut data)
```

Comment:
This still fails when we don't have any justification.

Reply:
if there is no justification why is block valid ? in which situation does this happens, if it happens when you import your own block then we should check that it is indeed your block so you can trust it.

Is there another point in the code that check that the justification exist and I'm not aware of ? because if none I don't understand why this is valid to validate a block without justification.

Reply:
We will not persist the justifications, so when you are doing an initial sync, you will get the blocks without any justification.

Reply:
I'll have to understand where unjustified new block import will be refused but at least this is done now